include/seahorn/Expr/ExprRewriter.hh

```diff
@@ -0,0 +1,144 @@
+#pragma once
+#include "seahorn/Expr/Expr.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprRewriteRule.hh"
+
+namespace expr {
+using namespace seahorn;
+
+Expr rewriteITEComp(Expr exp);
+
+struct RewriteFrame {
+  Expr m_exp;
+  size_t m_depth;
+  size_t m_i; // up to m_i th children have been rewritten
+  RewriteFrame(Expr exp, size_t depth, size_t i)
+      : m_exp(exp), m_depth(depth), m_i(i) {}
+};
+
+using ExprVisitedMap = std::unordered_map<ENode *, bool>;
+using RewriteFrameVector = std::vector<RewriteFrame>;
+
+class ExprRewriterConfig {
+  /* apply rewrite rules */
+protected:
+  ExprFactory &m_efac; // for making expr
+  EZ3 &m_zctx;         // for z3 simplifier if needed
```

Comment:
I don't see why generic rewriter should depend on Z3. Its specialization can, but the generic rewriter should not